src/gui/editors/SongEditor.cpp

```diff
@@ -52,26 +52,84 @@
 #include "PianoRoll.h"
 #include "Track.h"
 
-positionLine::positionLine( QWidget * parent ) :
-	QWidget( parent )
+const QVector<double> positionLine::s_zoomLevels =
+		{ 0.125f, 0.25f, 0.5f, 1.0f, 2.0f, 4.0f, 8.0f, 16.0f };
+
+positionLine::positionLine( QWidget* parent, ComboBoxModel* zoom) :
+	QWidget( parent ),
+	m_tailGradient ( false ),
+	m_lineColor (0, 0, 0, 0)
 {
-	setFixedWidth( 1 );
+	m_currentZoom = zoom;
+	resize( 8, height() );
+	
 	setAttribute( Qt::WA_NoSystemBackground, true );
+	setAttribute( Qt::WA_TransparentForMouseEvents );
 }
 
-
-
-
-void positionLine::paintEvent( QPaintEvent * pe )
+void positionLine::paintEvent( QPaintEvent* pe )
 {
 	QPainter p( this );
-	p.fillRect( rect(), QColor( 255, 255, 255, 153 ) );
+	
+	// Resize based on the zoom value
+	resize( 8.0f * s_zoomLevels[ m_currentZoom->value() ], height() );
+	
+	// If width is 1, we don't need a gradient
+	if (width() == 1)
+	{
+		p.fillRect( rect(),
+			QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 153) );
+	}
+	
+	// If width > 1, we need the gradient
+	else
+	{
+		// Create the gradient trail behind the line
+		QLinearGradient gradient( rect().bottomLeft(), rect().bottomRight() );
+		
+		// If gradient is enabled, we're in focus and we're playing, enable gradient
+		if (Engine::getSong()->isPlaying() && m_tailGradient && 
+			Engine::getSong()->playMode() == Song::Mode_PlaySong)
+		{
+			gradient.setColorAt((double)( ( width() - 1.0f )/width() ),
+				QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 60) );
+		}
+		else
+		{
+			gradient.setColorAt((double)( ( width() - 1.0f )/width() ),
+				QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 0) );
+		}
+		
+		// Fill in the remaining parts
+		gradient.setColorAt(0,
+			QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 0) );
+		gradient.setColorAt(1,
+			QColor( m_lineColor.red(), m_lineColor.green(), m_lineColor.blue(), 153) );
```

Comment:
Sorry, I forgot to mention this in the previous review. Did you intend to switch to using the suggestion I made before?
```cpp
		// Create the gradient trail behind the line
		QLinearGradient gradient(rect().bottomLeft(), rect().bottomRight());
		auto lineColor = m_lineColor;

		// If gradient is enabled, we're in focus and we're playing, enable gradient
		const auto enableGradient = Engine::getSong()->isPlaying() && m_tailGradient
			&& Engine::getSong()->playMode() == Song::Mode_PlaySong;
		lineColor.setAlpha(enableGradient ? 60 : 0);
		gradient.setColorAt((width() - 1.0) / width(), lineColor);

		// Fill in the remaining parts
		lineColor.setAlpha(0);
		gradient.setColorAt(0, lineColor);
		lineColor.setAlpha(153);
		gradient.setColorAt(1, lineColor);
```

Reply:
I actually decided not to do this because although my code is larger, it’s easily understood and doesn’t involve creating new variables. Your code works just as well but I just thought the original code is more friendly to people new to the code. If you do want me to implement this however, I’d be glad to.

Reply:
I don't mind either way - I just wondered whether you'd forgotten since you said you were going to change it. If you prefer the old way, let's go with that.

That said, would you mind removing the `double` casts? C-style casts are usually avoided in C++, and since `float` can implicitly convert to `double`, no cast is needed anyway. You could even change `1.0f` to `1.0`, so the calculation only uses `double`s in the first place.

Reply:
Yep, I've removed the `double` casts.